### PR TITLE
[feature]resize pvc

### PR DESCRIPTION
**Pull request by jiuker**
// 1.Get the data already stored internally
	// 2.Get the desired data
	// 3.Start querying the pvc list when you find data inconsistencies
	// 3.1 Comparison using real pvc capacity and desired data
	// 3.1.1 Update if you find inconsistencies
	// 3.2 Writing successful updates to internal
	// 4. Set to old VolumeClaimTemplates to update.Prevent update error reporting
	// 5. Set to old annotations to update
When pvc is too small,we can use it to make it bigger.
Fixes # .

Changes proposed on the PR:
- Make pvc size bigger. When it need.

### CODE REVIEW

**Comment by jiuker** [comment]
@ese

**Comment by jiuker** [comment]
@ese Any suggestions?

**Comment by samof76** [comment]
@jiuker this is good feature... but should have some integration tests for this?

**Comment by jiuker** [comment]
@samof76 yeah.review plz

File: service/k8s/statefulset.go [hunk 1]
```diff
@@ -107,6 +110,68 @@ func (s *StatefulSetService) CreateOrUpdateStatefulSet(namespace string, statefu
...
+			updateFailed := false
+			realUpdate := false
+			for _, pvc := range pvcs.Items {
+				realCapacity := pvc.Spec.Resources.Requests.Storage().Value()
+				if realCapacity != stateCapacity {
+					realUpdate = true
+					pvc.Spec.Resources.Requests = statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests
+					_, err = s.kubeClient.CoreV1().PersistentVolumeClaims(storedStatefulSet.Namespace).Update(context.Background(), &pvc, metav1.UpdateOptions{})
+					if err != nil {
+						if !updateFailed {
```
Comment (ese) [inline service/k8s/statefulset.go:155]: I think this if statement is not needed since updateFailed will be always `false` here

**Comment by ese** [comment]
I will do a deep review later. Sorry for the delay but I was not confident with resizing workflow.

File: service/k8s/statefulset.go [hunk 2]
```diff
@@ -107,6 +110,68 @@ func (s *StatefulSetService) CreateOrUpdateStatefulSet(namespace string, statefu
...
+						s.logger.WithField("namespace", namespace).WithField("pvc", rfName).Warningf("resize pvc failed:%s", err.Error())
+					}
+				}
+			}
+			if !updateFailed && len(pvcs.Items) != 0 {
+				annotations["storageCapacity"] = fmt.Sprintf("%d", stateCapacity)
+				storedStatefulSet.Annotations = annotations
+				if realUpdate {
+					s.logger.WithField("namespace", namespace).WithField("pvc", rfName).Infof("resize pvc will resize from %d to %d Success", storedCapacity, stateCapacity)
+				} else {
```
Comment (ese) [inline service/k8s/statefulset.go:167]: I would remove these else stament not logging at all for the first annotation set

File: service/k8s/statefulset.go [hunk 3]
```diff
@@ -107,6 +110,68 @@ func (s *StatefulSetService) CreateOrUpdateStatefulSet(namespace string, statefu
...
+					s.logger.WithField("namespace", namespace).WithField("pvc", rfName).Warningf("set annotations,resize nothing")
+				}
```
Comment (ese) [inline service/k8s/statefulset.go:169]: I would remove this else stament not logging at all for the first annotation set

File: service/k8s/statefulset.go [hunk 4]
```diff
@@ -107,6 +110,68 @@ func (s *StatefulSetService) CreateOrUpdateStatefulSet(namespace string, statefu
...
+				realCapacity := pvc.Spec.Resources.Requests.Storage().Value()
+				if realCapacity != stateCapacity {
+					realUpdate = true
+					pvc.Spec.Resources.Requests = statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests
+					_, err = s.kubeClient.CoreV1().PersistentVolumeClaims(storedStatefulSet.Namespace).Update(context.Background(), &pvc, metav1.UpdateOptions{})
+					if err != nil {
+						if !updateFailed {
+							updateFailed = true
+						}
+						s.logger.WithField("namespace", namespace).WithField("pvc", rfName).Warningf("resize pvc failed:%s", err.Error())
```
Comment (ese) [inline service/k8s/statefulset.go:158]:
```suggestion
						s.logger.WithField("namespace", namespace).WithField("pvc", pvc.Name).Warningf("resize pvc failed:%s", err.Error())
```

**Comment by ese** [comment]
Thanks @jiuker.

File: service/k8s/statefulset.go (hunk 1 above)
Comment (jiuker) [reply service/k8s/statefulset.go:155]: Not supporting the update is only one of the reasons for the failure of the update, and the network is also possible. Or the current PVC has been artificially modified

Comment (ese) [reply service/k8s/statefulset.go:155]: I mean, we can replcae

```
if !updateFailed {
  updateFailed=true
}
```
by simply
```
  updateFailed=true
```
We don't need do that check to set failed to true

File: service/k8s/statefulset.go (hunk 3 above)
Comment (ese) [reply service/k8s/statefulset.go:169]: Why did you mark as resolved? It would help know your opinion here

**Review by ese — APPROVED** [review]

**Pull request merged**